### PR TITLE
Encapsulation of applier logic: Move Txn related code out of applier.go.

### PR DESCRIPTION
**Pull request by ptabor**
The PR removes calls to applierV3base logic from server.go that is NOT part of 'application'.
The original idea was that read-only transaction and Range call shared logic with Apply,
so they can call appliers directly (but bypassing all 'corrupt', 'quota' and 'auth' wrappers).

This PR moves all the logic to a separate file (that later can become package on its own).

This is preparation to fix: #13766 by wrapping backends.

### CODE REVIEW

File: server/etcdserver/server.go [hunk 1]
```diff
@@ -1050,6 +1048,10 @@ func (s *EtcdServer) applySnapshot(ep *etcdProgress, apply *apply) {
 	ep.appliedi = apply.snapshot.Metadata.Index
 	ep.snapi = ep.appliedi
 	ep.confState = apply.snapshot.Metadata.ConfState
+
+	// As backends and implementations like alarmsStore changed, we need
+	// to re-bootstrap Appliers.
+	s.uberApply = newUberApplier(s)
```
Comment (serathius) [inline server/etcdserver/server.go]: Instead of storing a alarmStore reference, could we store reference to alarmStore getter interface?

File: server/etcdserver/server.go [hunk 2]
```diff
@@ -1051,7 +1051,12 @@ func (s *EtcdServer) applySnapshot(ep *etcdProgress, apply *apply) {
 
 	// As backends and implementations like alarmsStore changed, we need
 	// to re-bootstrap Appliers.
-	s.uberApply = newUberApplier(s)
+	s.uberApply = s.NewUberApplier()
```
Comment (serathius) [inline server/etcdserver/server.go:1056]: This is even more fragile then before, can we remove some of those fields from server?

File: server/etcdserver/api/v3rpc/header.go [hunk 3]
```diff
@@ -17,12 +17,13 @@ package v3rpc
 import (
 	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
 	"go.etcd.io/etcd/server/v3/etcdserver"
+	apply2 "go.etcd.io/etcd/server/v3/etcdserver/apply"
```
Comment (serathius) [inline server/etcdserver/api/v3rpc/header.go:20]: Just `apply`?

File: server/etcdserver/api/v3rpc/watch.go [hunk 4]
```diff
@@ -26,6 +26,7 @@ import (
 	"go.etcd.io/etcd/api/v3/v3rpc/rpctypes"
 	"go.etcd.io/etcd/server/v3/auth"
 	"go.etcd.io/etcd/server/v3/etcdserver"
+	apply2 "go.etcd.io/etcd/server/v3/etcdserver/apply"
```
Comment (serathius) [inline server/etcdserver/api/v3rpc/watch.go:29]: ditto

File: server/etcdserver/apply/uber_applier.go [hunk 5]
```diff
@@ -103,21 +102,22 @@ func (a *uberApplier) RestoreAlarms() {
...
+	ar := &ApplyResult{}
 	defer func(start time.Time) {
-		success := ar.err == nil || ar.err == mvcc.ErrCompacted
-		applySec.WithLabelValues(v3Version, op, strconv.FormatBool(success)).Observe(time.Since(start).Seconds())
-		warnOfExpensiveRequest(a.lg, a.warningApplyDuration, start, &pb.InternalRaftStringer{Request: r}, ar.resp, ar.err)
-		if !success {
-			warnOfFailedRequest(a.lg, start, &pb.InternalRaftStringer{Request: r}, ar.resp, ar.err)
-		}
+		op += " "
+		//		success := ar.Err == nil || ar.Err == mvcc.ErrCompacted
```
Comment (serathius) [inline server/etcdserver/apply/uber_applier.go]: TODO?

File: server/etcdserver/server.go (hunk 2 above)
Comment (ptabor) [reply server/etcdserver/server.go:1056]: It's not worse than what used to happen. 

We called restoreAlarms:
https://github.com/etcd-io/etcd/blob/a08d479463960948d5d9c3508623ade915c6e66f/server/etcdserver/server.go#L1003

That underneath were reconstructing the appliers and wrapping them:
https://github.com/etcd-io/etcd/blob/a08d479463960948d5d9c3508623ade915c6e66f/server/etcdserver/server.go#L2239-L2253

I'm afraid there are no fields to remove

File: server/etcdserver/server.go (hunk 1 above)
Comment (ptabor) [reply server/etcdserver/server.go]: I would agree in general, but it would require deep reform what `Backend` is and how restore from snapshot works. 

Currently restore from snapshot: 
  - creates a new instance of the backend
  - all the objects/helpers pointing on the all backend are 'poinsoned'. 
  - we need to rebootstrap them using the new backend

It might be easier / safer to create them from scratch rather make sure there is `foo.RebootstrapFromBackend(BE)` method that will entirely clean it's state and will load what needed from `be`

File: server/etcdserver/api/v3rpc/header.go (hunk 3 above)
Comment (ptabor) [reply server/etcdserver/api/v3rpc/header.go:20]: Done.

File: server/etcdserver/api/v3rpc/watch.go (hunk 4 above)
Comment (ptabor) [reply server/etcdserver/api/v3rpc/watch.go:29]: Done.

File: server/etcdserver/apply/uber_applier.go (hunk 5 above)
Comment (ptabor) [reply server/etcdserver/apply/uber_applier.go]: Sorry - seems I didn't exported the newest version. PTAL

**Comment by ptabor** [comment]
```
Sending build context to Docker daemon  84.81MB
Sending build context to Docker daemon  86.13MB


Step 1/14 : FROM ubuntu:21.10

toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
```

**Comment by ahrtr** [comment]
Is this PR planned on 3.6 or 3.7? I assume it's 3.7, otherwise it will be risky to cherry pick to 3.5.

**Comment by ptabor** [comment]
> Is this PR planned on 3.6 or 3.7? I assume it's 3.7, otherwise it will be risky to cherry pick to 3.5.

I assumed 3.6 - it's early in the release train I think. 
I have doubts about 3.5 - as it's pretty substantial in volume change for patch release. 

> Are you going to move all read-only transactions out of the normal raft consensus protocol (apply workflow), and instead depends on [linearizableReadLoop](https://github.com/etcd-io/etcd/blob/a08d479463960948d5d9c3508623ade915c6e66f/server/etcdserver/v3_server.go#L713) in the next step? In this case, the consistent_index will only be moved forward by the mvcc transactions?
> 
> It seems that there are still some big refactors to completely resolve [issues/13766](https://github.com/etcd-io/etcd/issues/13766).

No. These transactions (RO through apply) are not harmful as they are fully idempotent. 

The idea was to split the serving stack into 




Currently: 
image

Splitted stack serving 'applies' and read-only calls:
image

Practically needed stack due to performance reasons:
image

But we need to be careful that any 'async' calls will not influence cache content such that it can meaningful impact on the 'apply' determinism. 

But all of this seems indeed complex for 3.6. 

So in short-term I'm thinking about:  

image

That makes a sandwich around existing logic, with verification of origin of the transaction.

**Comment by ahrtr** [comment]
> I assumed 3.6 - it's early in the release train I think.
> I have doubts about 3.5 - as it's pretty substantial in volume change for patch release.

If it's planned for 3.6, then my PR [13854](https://github.com/etcd-io/etcd/pull/13854) is definitely not needed any more, at least not needed for 3.6.  What's the plan for 3.5? 


> So in short-term I'm thinking about:

I do not get time to review through this PR (will probably had a deep review tomorrow), but based on my quick review this morning, it seems that this PR has already completed this short-term, correct?

**Comment by serathius** [comment]
@ahrtr PTAL https://github.com/etcd-io/etcd/pull/13854#issuecomment-1088623477

**Comment by serathius** [comment]
Working on the postmortem https://github.com/etcd-io/etcd/pull/13967 made me in different direction than @ptabor. I think that overall idea of using consistent_index fields on EtcdServer and using hooks on backend to flush it is wrong. Will think on this more and maybe send a draft.

**Comment by ptabor** [comment]
>  I think that overall idea of using consistent_index fields on EtcdServer and using hooks on backend to flush it is wrong

Putting on side future directions, do you think that this refactoring is not worthwhile on its own ? 
 - splitting enormous etc-server package
 - reducing number of applier layers directly exposed in EtcdServer object 
 - avoiding recursive logic of applier's calling back to server to call nested applier's again
 - splitting post-raft code (appliers) from code that is sometimes called as part of application, and sometimes on its own (txn).

**Comment by serathius** [comment]
You're right, it's not an alternative, but supplementary improvement. We would greatly benefit from it.

File: server/etcdserver/apply/apply.go [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+	raftStatus      RaftStatusGetter
+	snapshotServer  SnapshotServer
+	consistentIndex cindex.ConsistentIndexer
+
+	txnModeWriteWithSharedBuffer bool
+}
+
+func newApplierV3Backend(
+	lg *zap.Logger,
+	kv mvcc.KV,
```
Comment (ahrtr) [inline server/etcdserver/apply/apply.go:132]: mvcc.WatchableKV?

File: server/etcdserver/apply/apply.go [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+	DowngradeInfoSet(r *membershippb.DowngradeInfoSetRequest, shouldApplyV3 membership.ShouldApplyV3)
+}
+
+type SnapshotServer interface {
+	ForceSnapshot()
+}
+
+type applierV3backend struct {
+	lg              *zap.Logger
+	kv              mvcc.KV
```
Comment (ahrtr) [inline server/etcdserver/apply/apply.go:118]: mvcc.WatchableKV?

File: server/etcdserver/apply/uber_applier.go [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+		a.applyV3 = newApplierV3Capped(a.applyV3)
+	}
+	if corruptAlarms {
+		a.applyV3 = newApplierV3Corrupt(a.applyV3)
+	}
+}
+
+func (a *UberApplier) Apply(r *pb.InternalRaftRequest, shouldApplyV3 membership.ShouldApplyV3) *ApplyResult {
+	// We first execute chain of WrapApply across all objects (e.g. CorruptApplier -> CappedApplier -> Auth -> Quota -> Backend),
+	// than dispatch(), than individual methods wrappers CorruptApplier.Put(CappedApplier.Put(... BackendApplier.Put())))
```
Comment (ahrtr) [inline server/etcdserver/apply/uber_applier.go:108]: typo, such as "than" --> "then". 

Please also try to refine the comment if possible.

File: server/etcdserver/apply/uber_applier.go [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+func (a *UberApplier) RestoreAlarms() {
+	noSpaceAlarms := len(a.alarmStore.Get(pb.AlarmType_NOSPACE)) > 0
+	corruptAlarms := len(a.alarmStore.Get(pb.AlarmType_CORRUPT)) > 0
+	a.applyV3 = a.applyV3base
+	if noSpaceAlarms {
+		a.applyV3 = newApplierV3Capped(a.applyV3)
+	}
+	if corruptAlarms {
+		a.applyV3 = newApplierV3Corrupt(a.applyV3)
+	}
```
Comment (ahrtr) [inline server/etcdserver/apply/uber_applier.go:103]: Can both `noSpaceAlarms` and `corruptAlarms` be `true`?  If not, then please add a comment to clarify or update the code to avoid confusion.

File: server/etcdserver/apply/uber_applier.go [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+	"go.etcd.io/etcd/server/v3/etcdserver/api/v3alarm"
+	"go.etcd.io/etcd/server/v3/etcdserver/cindex"
+	"go.etcd.io/etcd/server/v3/etcdserver/txn"
+	"go.etcd.io/etcd/server/v3/lease"
+	"go.etcd.io/etcd/server/v3/storage/backend"
+	"go.etcd.io/etcd/server/v3/storage/mvcc"
+	"go.uber.org/zap"
+)
+
+type UberApplier struct {
```
Comment (ahrtr) [inline server/etcdserver/apply/uber_applier.go:33]: Is it possible to define an interface for `UberApplier`? In this case, `Etcdserver` just holds a reference to the instance of the interface.

File: server/etcdserver/apply/uber_applier.go [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+	}
+	if corruptAlarms {
+		a.applyV3 = newApplierV3Corrupt(a.applyV3)
+	}
+}
+
+func (a *UberApplier) Apply(r *pb.InternalRaftRequest, shouldApplyV3 membership.ShouldApplyV3) *ApplyResult {
+	// We first execute chain of WrapApply across all objects (e.g. CorruptApplier -> CappedApplier -> Auth -> Quota -> Backend),
+	// than dispatch(), than individual methods wrappers CorruptApplier.Put(CappedApplier.Put(... BackendApplier.Put())))
+	return a.applyV3.WrapApply(context.TODO(), r, shouldApplyV3, a.dispatch)
```
Comment (ahrtr) [inline server/etcdserver/apply/uber_applier.go:109]: Why do yo introduce another layer of encapsulation/dispatch? I'd suggest to do it in the following PR, so that we can keep this PR as simple/clean as possible.

File: server/etcdserver/txn/txn.go [hunk 12]
```diff
@@ -0,0 +1,682 @@
...
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	"go.etcd.io/etcd/pkg/v3/traceutil"
+	"go.etcd.io/etcd/server/v3/auth"
+	"go.etcd.io/etcd/server/v3/etcdserver/etcderrors"
+	"go.etcd.io/etcd/server/v3/lease"
+	"go.etcd.io/etcd/server/v3/storage/mvcc"
+	"go.uber.org/zap"
+)
+
+func Put(ctx context.Context, lg *zap.Logger, lessor lease.Lessor, kv mvcc.KV, txn mvcc.TxnWrite, p *pb.PutRequest) (resp *pb.PutResponse, trace *traceutil.Trace, err error) {
```
Comment (ahrtr) [inline server/etcdserver/txn/txn.go:32]: nit: Please rename the `txn`, so as not to be the same as the package name .

File: server/etcdserver/apply/uber_applier.go (hunk 11 above)
Comment (ptabor) [reply server/etcdserver/apply/uber_applier.go:109]: It's not another layer, it's just making an explicit what was subtle & hidden so far. 

Please see this comment: 
https://github.com/etcd-io/etcd/blob/d849269dab5c07c1259de4441809d41135e7d0d8/server/etcdserver/apply.go#L166-L170

The call down the hierarchy was first traversing the recursive 'Apply`call structure, to get translated to a specific method, to traverse the overrides of the specific methods down the hierarchy. 

I broke the 'cyclic' dependency without changing the call hierarchy.

File: server/etcdserver/server.go [hunk 13]
```diff
@@ -33,7 +33,10 @@ import (
 	humanize "github.com/dustin/go-humanize"
 	"github.com/prometheus/client_golang/prometheus"
 	"go.etcd.io/etcd/pkg/v3/notify"
+	"go.etcd.io/etcd/pkg/v3/runtime"
 	"go.etcd.io/etcd/server/v3/config"
+	apply2 "go.etcd.io/etcd/server/v3/etcdserver/apply"
```
[the remark this answers is not in the document]
Comment (ptabor) [reply server/etcdserver/server.go:38]: It required renames of the 'apply' struct in the package... but I did that.

File: server/etcdserver/etcderrors/errors.go [hunk 14]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package etcdserver
+package etcderrors
```
[the remark this answers is not in the document]
Comment (ptabor) [reply server/etcdserver/etcderrors/errors.go:15]: Done.

File: server/etcdserver/apply/apply.go (hunk 6 above)
Comment (ptabor) [reply server/etcdserver/apply/apply.go:132]: If the more narrow interface is sufficient, I think we should use the more narrow.

File: server/etcdserver/apply/uber_applier.go (hunk 8 above)
Comment (ptabor) [reply server/etcdserver/apply/uber_applier.go:108]: PTAL

File: server/etcdserver/apply/uber_applier.go (hunk 9 above)
Comment (ptabor) [reply server/etcdserver/apply/uber_applier.go:103]: Unlikely but theoretically possible. This preserves the original logic: 

https://github.com/etcd-io/etcd/blob/52ea811af89ae3275b69054e815fa7f9a0471240/server/etcdserver/server.go#L2274-L2279

File: server/etcdserver/apply/uber_applier.go (hunk 10 above)
Comment (ptabor) [reply server/etcdserver/apply/uber_applier.go:33]: Done

File: server/etcdserver/txn/txn.go (hunk 12 above)
Comment (ptabor) [reply server/etcdserver/txn/txn.go:32]: Done

File: server/etcdserver/api/v3rpc/quota.go [hunk 15]
```diff
@@ -86,6 +86,10 @@ func (s *quotaLeaseServer) LeaseGrant(ctx context.Context, cr *pb.LeaseGrantRequ
 func NewQuotaLeaseServer(s *etcdserver.EtcdServer) pb.LeaseServer {
 	return &quotaLeaseServer{
 		NewLeaseServer(s),
-		quotaAlarmer{storage.NewBackendQuota(s.Cfg, s.Backend(), "lease"), s, s.ID()},
+		quotaAlarmer{newBackendQuota(s, "lease"), s, s.MemberId()},
 	}
 }
+
+func newBackendQuota(s *etcdserver.EtcdServer, name string) storage.Quota {
+	return storage.NewBackendQuota(s.Logger(), s.Cfg.QuotaBackendBytes, s.Backend(), "lease")
```
[the remark this answers is not in the document]
Comment (ahrtr) [reply server/etcdserver/api/v3rpc/quota.go:94]: The "lease" might be a typo, it should be the input parameter `name`. Please update this.

[the remark this answers is not in the document]
Comment (ptabor) [reply server/etcdserver/api/v3rpc/quota.go]: Done

File: server/etcdserver/apply/apply.go (hunk 6 above)
Comment (ahrtr) [reply server/etcdserver/apply/apply.go:132]: agreed.

File: server/etcdserver/apply/uber_applier.go (hunk 9 above)
Comment (ahrtr) [reply server/etcdserver/apply/uber_applier.go:107]: Probably we should add a verification (only one kind of alarm can exist) here in case whatever potential bug or error.

File: server/etcdserver/server.go (hunk 1 above)
Comment (ahrtr) [reply server/etcdserver/server.go]: I think we can revisit this comment in a separate PR if needed.

File: server/etcdserver/apply/uber_applier.go [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+func (a *uberApplier) Apply(r *pb.InternalRaftRequest, shouldApplyV3 membership.ShouldApplyV3) *ApplyResult {
+	// We first execute chain of WrapApply() calls down the hierarchy:
+	// (i.e. CorruptApplier -> CappedApplier -> Auth -> Quota -> Backend),
+	// then dispatch() unpacks the request to a specific method (like Put),
+	// that gets executed down the hierarchy again:
+	// i.e. CorruptApplier.Put(CappedApplier.Put(...(BackendApplier.Put(...)))).
+	return a.applyV3.WrapApply(context.TODO(), r, shouldApplyV3, a.dispatch)
+}
+
+// This function
```
[the remark this answers is not in the document]
Comment (ptabor) [reply server/etcdserver/apply/uber_applier.go]: Thank you for noticing. Fixed.

File: server/etcdserver/txn/util.go [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+//
+//func (n panicAlternativeStringer) String() (s string) {
+//	defer func() {
+//		if err := recover(); err != nil {
+//			s = n.alternative()
+//		}
+//	}()
+//	s = n.stringer.String()
+//	return s
+//}
```
Comment (ahrtr) [inline server/etcdserver/txn/util.go:122]: Do you keep the comment intentionally?  If not, then I'd suggest to remove them.

File: server/etcdserver/apply/uber_applier.go (hunk 11 above)
Comment (ptabor) [reply server/etcdserver/apply/uber_applier.go]: 1. Renamed - but to be honest I don't think it 'less' misleading.  
The function does not full 'Apply' it does only provides generic pre & post apply logic... and the actual application is delegated to the 'applyFunc' provided as a parameter.  
I have no better idea how to get rid of 'applyFunc'. 

Intuitively such stack of executions would be IMHO the most expected:
```
uberApplier.Apply:
  dispatcher()
    authApplierV3.Put
       authApplierV3.Pre
       quotaApplierV3.Put
           quotaApplierV3.Pre
           applierV3backend.Put
             applierV3backend.Pre
             (actual application)
             applierV3backend.Post
           quotaApplierV3.Post   
       authApplierV3.Post
 ```
 
 But I don't see how to implement it generically in go... in not error-prone way. 
 
 2. I unexported `ApplyFunc` type as it's indeed a local. 
    All the interfaces and implementations are 'private' (i.e. lower case names). 
    I have doubts about lower-case methods within the interfaces and classed. 
    Maybe I'm too biased in Java style of thinking, but in Java the interface methods are always public...
    but the interface itself can be package local.   I think that lower case methods of class are class private implementation detail, while the Capital methods are expected to be called by other classes within the same packages. I know it's not the golang specification... but also I don't see here risk that anyone would use the methods from outside of the package... as only exported think is `UberApplier` with a very narrow interface.

File: server/etcdserver/txn/util.go (hunk 17 above)
Comment (ptabor) [reply server/etcdserver/txn/util.go]: Fixed. Don't remember the origin.

File: server/etcdserver/apply/uber_applier.go (hunk 9 above)
Comment (ptabor) [reply server/etcdserver/apply/uber_applier.go:107]: I don't see why only 1 alarm is an invariant of the system. 
What if we are in NOSPACE alarm and the corruption check will declare the in-consistency ? 
I assume the applier will have 2 wrappers, but everything should work as expected.

File: server/etcdserver/apply/uber_applier.go (hunk 11 above)
Comment (ahrtr) [reply server/etcdserver/apply/uber_applier.go]: This PR is already too big, so Let's discuss & address this in a separate session. The naming and structure looks much better now.

File: server/etcdserver/apply/uber_applier.go (hunk 9 above)
Comment (ahrtr) [reply server/etcdserver/apply/uber_applier.go:107]: You do not change it in this PR, Let's keep it as it is for now. We can address it separately if needed.

**Comment by ahrtr** [comment]
This is a huge PR, which in general I don't think it's good for review. It took me huge time to review this PR, although I still don't think it's well reviewed, but overall it looks good to me. 

If I do this PR, I will breakdown it to three PRs at least.  The first PR just moves all apply related code into package/folder `apply`. The second PR move transaction related code into `txn`. The last PR will add some interface and refine the code. 

In summary, I'd like to let this PR to go, reasons:
1. The pipeline are all green (minimal requirement);
2. I have already reviewed this PR, but would be better if @serathius can double check;
3. It has big impact on other PR, so I'd like this PR to be merged soon.

**Review by ahrtr — APPROVED** [review]

File: server/etcdserver/server.go [hunk 18]
```diff
@@ -1072,7 +1072,12 @@ func (s *EtcdServer) applySnapshot(ep *etcdProgress, apply *apply) {
 
 	// As backends and implementations like alarmsStore changed, we need
 	// to re-bootstrap Appliers.
-	s.uberApply = newUberApplier(s)
+	s.uberApply = s.NewUberApplier()
+}
+
+func (s *EtcdServer) NewUberApplier() *uberApplier {
+	return newUberApplier(s.lg, s.be, s.KV(), s.alarmStore, s.authStore, s.lessor, s.cluster, s, s, s.consistIndex,
```
Comment (serathius) [inline server/etcdserver/server.go]: Are we sure that during runtime (after bootstrap) none of those fields on server are recreated? For example restoring snapshot, recovering auth etcd?

Comment (ptabor) [reply server/etcdserver/server.go]: They are in the `applySnapshot` and this is handled.  I haven't found other places were we 'recreate' them.

File: server/etcdserver/apply/apply_auth.go [hunk 19]
```diff
@@ -150,20 +151,8 @@ func checkTxnReqsPermission(as auth.AuthStore, ai *auth.AuthInfo, reqs []*pb.Req
...
-func CheckTxnAuth(as auth.AuthStore, ai *auth.AuthInfo, rt *pb.TxnRequest) error {
-	for _, c := range rt.Compare {
-		if err := as.IsRangePermitted(ai, c.Key, c.RangeEnd); err != nil {
-			return err
-		}
-	}
-	if err := checkTxnReqsPermission(as, ai, rt.Success); err != nil {
-		return err
-	}
-	return checkTxnReqsPermission(as, ai, rt.Failure)
```
Comment (serathius) [inline server/etcdserver/apply/apply_auth.go:162]: Note: checkTxnReqsPermission was not removed from this module, so we have duplicate. 
Will confirm if this is cleaned in later commits.

**Review by serathius — APPROVED** [review]
one small with function removal

**Comment by ptabor** [comment]
FTR: Flake due to port conflict: 

```
=== Failed
=== FAIL: integration TestAuthority/Size:_1,_Scenario:_"https://address[:port]" (0.00s)
    cluster.go:532: Creating listener with addr: 127.0.0.1:2131405814
    cluster.go:532: Creating listener with addr: 127.0.0.1:213[150](https://github.com/etcd-io/etcd/runs/6524932268?check_suite_focus=true#step:5:151)5814
    logger.go:130: 2022-05-20T12:46:04.608Z	INFO	m0	LISTEN GRPC	{"member": "m0", "grpcAddr": "127.0.0.1:33130", "m.Name": "m0", "workdir": "/tmp/TestAuthoritySize_1,_Scenario_httpsaddressport821414494/001"}
    cluster.go:724: listenGRPC FAILED: listen failed on grpc socket 127.0.0.1:33130 (listen tcp 127.0.0.1:33130: bind: address already in use)
    --- FAIL: TestAuthority/Size:_1,_Scenario:_"https://address[:port]" (0.00s)
```

**Comment by serathius** [comment]
LGTM

File: server/etcdserver/txn/txn.go [hunk 20]
```diff
@@ -0,0 +1,682 @@
...
+	ops := rt.Success
+	if txnPath[0] = applyCompares(rv, rt.Compare); !txnPath[0] {
+		ops = rt.Failure
+	}
+	for _, op := range ops {
+		tv, ok := op.Request.(*pb.RequestOp_RequestTxn)
+		if !ok || tv.RequestTxn == nil {
+			continue
+		}
+		txnPath = append(txnPath, compareToPath(rv, tv.RequestTxn)...)
```
Comment (tisonkun) [inline server/etcdserver/txn/txn.go:515]: What if the nested transaction's conditions become different after applying previous ops?

Comment (lavacat) [reply server/etcdserver/txn/txn.go:515]: @tisonkun interesting question. I think all (including nested) `compare` clauses are evaluated [before](https://github.com/etcd-io/etcd/blob/main/server/etcdserver/txn/txn.go#L242) `applyTxn`, so applying nested Txn won't affect results of `compare`

Comment (tisonkun) [reply server/etcdserver/txn/txn.go:515]: Yeah. That's how the current logic is implemented. But it seems not quite intuitive. Said in transaction, users can assume the requests are handled one by one, so the following requests should be able to see previous mutations.

etcd implements read after write, that if you write `put k1 v1` and `get k1` in one transaction, `get k1` sees `v1`, but it's not applied for compares.
